src/libappimage/desktop_integration/Thumbnailer.cpp

```diff
@@ -0,0 +1,156 @@
+            std::string canonicalPathMd5 = getCanonicalPathMd5(appImage.getPath());
+            bf::path normalThumbnailPath = getNormalThumbnailPath(canonicalPathMd5);
+            bf::path largeThumbnailPath = getLargeThumbnailPath(canonicalPathMd5);
+
+            bf::remove(normalThumbnailPath);
+            bf::remove(largeThumbnailPath);
+        }
+
+        void Thumbnailer::generateNormalSizeThumbnail(const std::string& canonicalPathMd5,
+                                                      std::vector<char>& normalIconData) const {
+
+            bf::path normalThumbnailPath = getNormalThumbnailPath(canonicalPathMd5);
+            bf::create_directories(normalThumbnailPath.parent_path());
+
+            try {
+                std::vector<uint8_t> iconData{normalIconData.begin(), normalIconData.end()};
+                IconHandle iconHandle(iconData);
+                iconHandle.setSize(128);
+                iconHandle.save(normalThumbnailPath.string(), "png");
+                return;
+            } catch (const IconHandleError&) {}
+
+            // It wasn't possible to generate a thumbnail, therefore the the icon will be written as it's
+            bf::ofstream out(normalThumbnailPath);
+            out.write(normalIconData.data(), normalIconData.size());
+        }
+
+        void Thumbnailer::generateLargeSizeThumbnail(const std::string& canonicalPathMd5,
+                                                     std::vector<char>& largeIconData) const {
+            bf::path largeThumbnailPath = getLargeThumbnailPath(canonicalPathMd5);
+            bf::create_directories(largeThumbnailPath.parent_path());
+
+            try {
+                std::vector<uint8_t> iconData{largeIconData.begin(), largeIconData.end()};
+                IconHandle iconHandle(iconData);
+                iconHandle.setSize(256);
+                iconHandle.save(largeThumbnailPath.string(), "png");
+                return;
+            } catch (const IconHandleError&) {}
+
+            // It wasn't possible to generate a thumbnail, therefore the the icon will be written as it's
+            bf::ofstream out(largeThumbnailPath);
+            out.write(largeIconData.data(), largeIconData.size());
+            out.close();
+        }
+
+
+        std::string Thumbnailer::getCanonicalPathMd5(const std::string& appImagePath) const {
+            auto canonicalAppImagePath = bf::weakly_canonical(appImagePath).string();
+            auto md5Raw = appimage_get_md5(canonicalAppImagePath.c_str());
+            std::string canonicalPathMd5 = md5Raw ?: "";
+
+            free(md5Raw);
+            return canonicalPathMd5;
+        }
+
+        bf::path Thumbnailer::getNormalThumbnailPath(const std::string& canonicalPathMd5) const {
+            bf::path xdgCacheHomePath(xdgCacheHome);
+
+            bf::path normalThumbnailPath =
+                xdgCacheHomePath / "thumbnails/normal" / (canonicalPathMd5 + ".png");
```

Comment:
The `thumbnails` part should be a constant IMO. Same goes for the `png` extension. You can then also add comments e.g., why we hardcode `.png` (even I regularly forget about it).

Reply:
Same goes for other methods similar to this in this file of course.

Reply:
In this case, the png extension is not hardcoded. It's a requirement from the thumbnails spec.

Constant properties created.

Reply:
> It's a requirement from the thumbnails spec.

We _do_ hardcode it. But now you confused me: is that PNG data that is written, or might e.g., SVG end up in files with a `.png` extension?

Reply:
We already support `svg` scaling and also png. As long as the input is in one of those formats it will be handled properly. Otherwise, we do save it as it's with the `png` extension.

Reply:
There should rather be an error/warning if the format cannot be worked with. These are the things I don't want to see any more in the code base.

Reply:
There is: https://github.com/AppImage/libappimage/blob/88255b11461c7da6853516fac61c65d83c81a052/src/libappimage/desktop_integration/Thumbnailer.cpp#L105